public/app/features/query/components/QueriesTab.tsx

```diff
@@ -96,85 +101,62 @@ export class QueriesTab extends PureComponent<Props, State> {
     this.setState({ data });
   }
 
-  findCurrentDataSource(dataSourceName: string | null = this.props.panel.datasource): DataSourceSelectItem {
+  findCurrentDataSource(dataSourceName: string | null): DataSourceSelectItem {
     return this.datasources.find(datasource => datasource.value === dataSourceName) || this.datasources[0];
   }
 
   onChangeDataSource = async (newDsItem: DataSourceSelectItem) => {
-    const { panel } = this.props;
+    let { queries } = this.props;
     const { dataSourceItem } = this.state;
 
     // switching to mixed
     if (newDsItem.meta.mixed) {
-      // Set the datasource on all targets
-      panel.targets.forEach(target => {
-        if (target.datasource !== ExpressionDatasourceID) {
-          target.datasource = panel.datasource;
-          if (!target.datasource) {
-            target.datasource = config.defaultDatasource;
+      for (const query of queries) {
+        if (query.datasource !== ExpressionDatasourceID) {
+          query.datasource = query.datasource;
+          if (!query.datasource) {
+            query.datasource = config.defaultDatasource;
```

Comment:
Is it OK to mutate query here? Shouldn't we first make a copy of the `queries` object?

Reply:
this was from the old code, so no change here. But yes this should be done immutably